### PR TITLE
find element by nspredicate string.

### PR DESCRIPTION
**Pull request by rafael-chavez**
## Change list

Please provide briefly described change list which are you going to propose. 
 Added NSPredicate string to the java-client to support xcuitest driver nspredicate element search.
## Types of changes

What types of changes are you proposing/introducing to Java client?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details

Please provide more details about changes if it is necessary. If there are new features you can provide code samples which show the way they
work and possible use cases. Also you can create [gists](https://gist.github.com) with pasted java code samples or put them here using markdown. 
About markdown please read [Mastering markdown](https://guides.github.com/features/mastering-markdown/) and [Writing on GitHub](https://help.github.com/categories/writing-on-github/)

### CODE REVIEW

**Comment by rafael-chavez** [comment]
@jlipps @Jonahss 
This is to support the NSPredicate string added to xcuitesting framework driver https://github.com/facebook/WebDriverAgent/pull/124. This wont hurt being in master, people using the old UIA framework just wont be able to use this support.

**Comment by rafael-chavez** [comment]
FYI @ericwengelking

**Comment by TikhomirovSergey** [comment]
@rafael-chavez 
It is cool! :+1: 
But there are some remarks:

- Please improve the PR description

- please add some test which show how the feature works.

File: src/main/java/io/appium/java_client/MobileBy.java [hunk 1]
```diff
@@ -51,7 +51,15 @@ public static By AccessibilityId(final String id) {
 
         return new ByAccessibilityId(id);
     }
-
+    
+    public static By IosNsPredicateString(final String id) {
```
Comment (TikhomirovSergey) [inline src/main/java/io/appium/java_client/MobileBy.java]: Why Id?

Comment (rafael-chavez) [inline src/main/java/io/appium/java_client/MobileBy.java]: sorry let me fix this, should be iOSPredicateString

**Comment by rafael-chavez** [comment]
@TikhomirovSergey added test and updated pr description.

File: src/main/java/io/appium/java_client/MobileBy.java [hunk 2]
```diff
@@ -51,7 +51,15 @@ public static By AccessibilityId(final String id) {
 
         return new ByAccessibilityId(id);
     }
-
+    
+    public static By IosNsPredicateString(final String iOSNsPredicateString) {
+    	if (iOSNsPredicateString == null) {
+    	  throw new IllegalArgumentException("Must supply an iOS NsPredicate String");
```
Comment (ericwengelking) [inline src/main/java/io/appium/java_client/MobileBy.java]: Since this is just a comment I would format it "NSPredicate" instead of "NsPredicate"

**Comment by ericwengelking** [comment]
lgtm thanks @rafael-chavez !

File: src/main/java/io/appium/java_client/ios/IOSDriver.java [hunk 3]
```diff
@@ -172,6 +173,24 @@ public RequiredElementType findElementByIosUIAutomation(String using)
...
         return (List<RequiredElementType>) findElements("-ios uiautomation", using);
     }
+    
+    /**
+     * @throws org.openqa.selenium.WebDriverException This method is not applicable with browser/webview UI.
+     */
+    @SuppressWarnings("unchecked") @Override
+    public RequiredElementType findElementByIosNsPredicate(String using)
+    throws WebDriverException {
+        return (RequiredElementType) findElement("predicate string", using);
```
Comment (jlipps) [inline src/main/java/io/appium/java_client/ios/IOSDriver.java]: I'd prefer for the locator strategy to be called `"-ios ns predicate"` or `"-ios predicate string"`. either way it must have a `-` in front of it (see `"-ios uiautomation"` for analogy above)

**Comment by TikhomirovSergey** [comment]
@rafael-chavez I'm going to check it and merge it if everything is ok.

**Comment by rafael-chavez** [comment]
@TikhomirovSergey can we ship this?

**Comment by TikhomirovSergey** [comment]
@rafael-chavez 
Sorry for the late response. I totally forgot about this PR.

Ok. I think that we are not able to merge this PR because:

- I'm not sure that Appium is supporting xcuitest driver. Please read roadmap. 
https://github.com/appium/appium/blob/51285550eadb907e5cf90931a0c72ec7bc4a004d/ROADMAP.md#current-roadmap-projects. @jlipps Please correct me if I'm wrong. When this supporting will be here then this PR will be merged.

- You've modified the test _IOSDriverTest.java_. This modification is invalid. CI is going to be there in near future. If CI was here now then this PR validation would be failed. Can you move your test to another class?

**Comment by rafael-chavez** [comment]
Appium works with appium-xcuitest-driver (I have it running locally), however it does not come with it out of the box. Which one were you referring to @TikhomirovSergey ?

Yes ill isolate the xcui driver test that i added to a XCUIDriverTest.java class.

**Comment by TikhomirovSergey** [comment]
@rafael-chavez 
Ok. Please isolate the test and show here how do you run it. Also I'll ask a question to appium team.

**Comment by bootstraponline** [comment]
Hi, I'm in favor of this change being merged. The documentation should clearly state it's limited to the XCUITest driver because the old UIAutomation JavaScript driver also supports predicates.

Once the test is moved to `XCUIDriverTest.java` then I think it'll be ready for merging.

**Comment by rafael-chavez** [comment]
@TikhomirovSergey isolated the test.

**Comment by TikhomirovSergey** [comment]
@rafael-chavez 
Please add a new ISOLATED test. 
...And you can provide additional documentation/description if there are many details. You can put it here: https://github.com/appium/java-client/tree/master/docs. I'll add the link to [WIKI](https://github.com/appium/java-client/wiki).

...And then I'll merge it

**Comment by rafael-chavez** [comment]
Moved the test to a separate file.

@TikhomirovSergey Is this for me to create a Readme file with instructions on how to install appium-xcuitest-driver?

**Comment by SrinivasanTarget** [comment]
@rafael-chavez Yes, you can create one here, https://github.com/appium/java-client/tree/master/docs

File: src/test/java/io/appium/java_client/ios/XCUIDriverTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.junit.Test;
+
+import io.appium.java_client.MobileBy;
+import org.junit.*;
+import org.openqa.selenium.Point;
+import org.openqa.selenium.ScreenOrientation;
+import org.openqa.selenium.html5.Location;
+	
+public class XCUIDriverTest extends BaseIOSTest {
```
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/ios/XCUIDriverTest.java]: Please add remove _extends BaseIOSTest _ and **@BeforeClass**/**@BeforeTest** and **@AfterClass**/**@After** methods. They should instantiate IOSDriver with specific capabilities, I think. May be it needs some specific server flags. 

This test won't work. It is using regular iOS UI Automation.

**Comment by TikhomirovSergey** [comment]
@rafael-chavez 
Well. Test is isolated but there are some more remarks. Please make it  clear.

**Comment by TikhomirovSergey** [comment]
@rafael-chavez 
also please squash commits.

File: src/test/java/io/appium/java_client/ios/XCUIDriverTest.java (hunk 4 above)
Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/ios/XCUIDriverTest.java]: Please add remove extends BaseIOSTest and add @BeforeClass/@BeforeTest and @AfterClass/@After methods. They should instantiate IOSDriver with specific capabilities, I think. May be it needs some specific server flags.

This test won't work. It is using regular iOS UI Automation.

Comment (TikhomirovSergey) [inline src/test/java/io/appium/java_client/ios/XCUIDriverTest.java]: >This test won't work. It is using regular iOS UI Automation.
I've read this document 
https://github.com/appium/appium-xcuitest-driver#usage

But anyway BaseIOSTest uses another app. This test won't find the target element.

**Comment by rafael-chavez** [comment]
@TikhomirovSergey got it working with xcuitest driver. Ill push documentation then squash my commits.

screen shot 2016-04-29 at 12 01 35 pm

**Comment by rafael-chavez** [comment]
@TikhomirovSergey @SrinivasanTarget  according to github admin users need to squash? 

https://help.github.com/articles/about-pull-request-merge-squashing/

**Comment by bootstraponline** [comment]
> @TikhomirovSergey @SrinivasanTarget according to github admin users need to squash?

admins can squash for you, but it's better to do it yourself.

```
$ git reset --soft head~19
$ git commit --amend
$ git push -f
```

**Comment by rafael-chavez** [comment]
Thanks @bootstraponline !!!!

**Comment by TikhomirovSergey** [comment]
@rafael-chavez 
Ok. I'm merging this PR. 
There is #378. Code style issues have been resolved there.
https://github.com/appium/java-client/blob/master/docs/Note-for-developers.md#code-style

**Comment by TikhomirovSergey** [comment]
Also https://github.com/appium/java-client/issues/379 has been opened.

**Comment by TikhomirovSergey** [comment]
#398

**Pull request merged**